src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -370,15 +370,61 @@ def deleteBlocks(self):
         """
         _deleteBlocks_
         Find deletable blocks, then decide if to delete based on:
-        Is there an active subscription for dataset or block ?
-          If yes => set deleted=2
-          If no => next check
         Has transfer to all destinations finished ?
-          If yes => request block deletion, approve request, set deleted=1
+          If yes => Delete rules associated with the block, set deleted=1
           If no => do nothing (check again next cycle)
         """
-        # FIXME: figure out the proper logic for rule block deletion
-        logging.info("Starting deleteBlocks methods --> IMPLEMENT-ME!!!")
+        #Get list of blocks that can be deleted
+        blockDict = self.findDeletableBlocks.execute(transaction=False)
+
+        if not blockDict:
+            logging.info("No blocks were found")
+            return
+
+        deletableBlocks = []
+        containerDict = {}
+		#Populate containerDict, assigning each block to its correspondant container
+        for blockName in blockDict:
+            containerList = self.rucio.listParentDIDs(blockName, scope=self.scope)
+            for cont in containerList:
+                if cont['name'] not in containerDict:
+                    containerDict[cont['name']] = {'blocks':[], 'rse':[], 'copies':0}
+                containerDict[cont['name']]['blocks'].append(blockName)
+
+        for contName in containerDict:
+            cont = containerDict[contName]
+            rules = self.rucio.listDataRules(contName, self.scope)
+            #Retrieve requested copies and RSEs
+            for rule in rules:
+                if rule['account'] == self.rucioAcct:
+                    cont['copies'] += rule['copies']
+                    cont['rse'].extend(self.rucio.evaluateRSEExpression(rule['rse_expression'], useCache=True, returnTape=True))
+
+            #Get RSE in which the block is available
+            availableRSEs = self.rucio.getReplicaInfoForBlocks(block=cont['blocks'])
+
+            for rses in availableRSEs:
+                #Count how many of the copies were requested by this account
+                actualCopies = len(set(rses['replica']) & set(cont['rse']))
+
+                #If enough requested copies are available, block can be deleted
+                if cont['copies'] <= actualCopies:
```

Comment:
This would delete blocks with 0 copies.
Can you please remind me again what is the logic we discussed here? Is it that the container needs to have at least one rule and that the block needs to be AVAILABLE on at least one RSE?

Reply:
We will need either 1 or 2 copies of each block (2 for most RAW data, 1 for other datatiers). I'm (maybe dangerously) assuming that all the necessary container-level rules have been created. If that's the case, cont['copies'] will be equal to the number of copies Tier0 needs.

Reply:
With phedex, we get the destination sites from `dbsbuffer_dataset_subscription.site`. Is there going to be an equivalent column for Rucio?

Reply:
it should remain the same. I'm planning to make changes as backward compatible as possible

Reply:
I'll keep using `dbsbuffer_dataset_subscription.site` then. I'll also add a protection and a warning message in case a particular container has no requested copies.